### ISSUE TITLE
`encodeURIComponent` not called on model's id when serializing into the URL

### ISSUE THREAD

**Issue by bantic** [emberjs/ember.js#11497]
If a model's id includes a slash ("/") character, ember will still serialize it as-is into a dynamic route's url, resulting in a broken url.

[Example jsbin](http://emberjs.jsbin.com/gigoli). Tested on Ember 1.13.2 and Canary.

Example:

```
// router
App.Router.map(function() {
  this.route('user', {path: '/users/:user_id'});
});

// template
{{#link-to 'user' 'abc/def'}}User with non-standard id{{/link-to}}
```

Following the link changes the url to "/users/abc/def", which, if the user reloads the page, results in the error: `Error: Assertion Failed: The URL '/users/abc/def' did not match any routes in your application`. It seems like the id should be serialized with `encodeURIComponent` so that the url changes to "/users/abc%2Fdef".

**Comment by rwjblue** [comment]
I believe that ember-data does something similar in buildURL (for posting to the API). 

Seems like a good change to me. Is it possible to rely on the current behavior? Do we need to make this a two step change (deprecate if we detect it may be an issue and then fix in a subsequent release)?

For now, you can use one of the new Ember.Helper.helper's to url-encode the argument (hopefully that can unblock you)...

**Comment by bantic** [comment]
Ok, great, I'll take a look into submitting a fix. Using the `serialize` and `deserialize` hooks on the route also seems to work fine as a workaround for now.

**Comment by bantic** [comment]
I've created a PR for the route-recognizer that correctly handles decoding URI segments when recognizing urls and encoding the segments when generating urls: https://github.com/tildeio/route-recognizer/pull/55

When that is merged I can create a PR for ember that will point at the new versions and fix this issue.

I did a little research and discovered that Rails's router does this decoding and encoding, as well. A rails app with the route "/users/:id" will route the url "/users/abc%2Fdef" to the `Users#show` route with `params[:id] = "abc/def"`. And likewise in rails if a `user` has id "abc/def", doing `redirect_to user` will go to the url "/users/abc%2Fdef".

**Comment by kayomarz** [comment]
I came across this in my app too (with Ember 2.4). The change mentioned by @bantic would be nice to have. I've used a helper to circumvent this.

**Comment by pixelhandler** [comment]
@bantic so the newest PR for this issue is here, https://github.com/tildeio/route-recognizer/pull/91 correct?

**Comment by bantic** [comment]
Yeah, that's correct.

**Comment by frunjik** [comment]
What is the status of this ?
Is there a workaround for handling special characters in id's?

**Comment by locks** [comment]
The route-recognizer PR is merged!
